From f812114144d67f02db3a3a2ae6de658cd0cf4cc5 Mon Sep 17 00:00:00 2001
From: "Guoniu.zhou" <guoniu.zhou@nxp.com>
Date: Fri, 27 May 2022 18:26:27 +0800
Subject: [PATCH 108/274] LF-6244-03: dma: pxp: config block size to 16x16 for
 primary path

Select block size to 16x16 pixel blocks for PXP primary path in order
to improve PXP performance

Signed-off-by: Guoniu.zhou <guoniu.zhou@nxp.com>
Reviewed-by: Robby Cai <robby.cai@nxp.com>
---
 drivers/dma/pxp/pxp_dma_v3.c | 33 ++++++++++++++++++++++-----------
 1 file changed, 22 insertions(+), 11 deletions(-)

diff --git a/drivers/dma/pxp/pxp_dma_v3.c b/drivers/dma/pxp/pxp_dma_v3.c
index 422d9efb9..8dbadd560 100644
--- a/drivers/dma/pxp/pxp_dma_v3.c
+++ b/drivers/dma/pxp/pxp_dma_v3.c
@@ -960,14 +960,14 @@ enum {
 };
 
 enum pxp_devtype {
-	PXP_V3,
+	PXP_V3 = 0,
 	PXP_V3P,	/* minor changes over V3, use WFE_B to replace WFE_A */
 	PXP_V3_8ULP,	/* PXP V3 version for iMX8ULP */
 };
 
-#define pxp_is_v3(pxp) ((pxp->devdata->version == 30) || \
-			(pxp->devdata->version == 32))
-#define pxp_is_v3p(pxp) (pxp->devdata->version == 31)
+#define pxp_is_v3(pxp) ((pxp->devdata->version == PXP_V3) || \
+			(pxp->devdata->version == PXP_V3_8ULP))
+#define pxp_is_v3p(pxp) (pxp->devdata->version == PXP_V3P)
 
 struct pxp_devdata {
 	void (*pxp_wfe_a_configure)(struct pxps *pxp);
@@ -991,7 +991,7 @@ static const struct pxp_devdata pxp_devdata[] = {
 		.pxp_dithering_configure = pxp_dithering_configure,
 		.pxp_data_path_config = NULL,
 		.pxp_restart = NULL,
-		.version = 30,
+		.version = PXP_V3,
 	},
 	[PXP_V3P] = {
 		.pxp_wfe_a_configure = pxp_wfe_a_configure_v3p,
@@ -1002,7 +1002,7 @@ static const struct pxp_devdata pxp_devdata[] = {
 		.pxp_dithering_configure = pxp_dithering_configure_v3p,
 		.pxp_data_path_config = pxp_data_path_config_v3p,
 		.pxp_restart = NULL,
-		.version = 31,
+		.version = PXP_V3P,
 	},
 	[PXP_V3_8ULP] = {
 		.pxp_wfe_a_configure = pxp_wfe_a_configure,
@@ -1013,7 +1013,7 @@ static const struct pxp_devdata pxp_devdata[] = {
 		.pxp_dithering_configure = pxp_dithering_configure,
 		.pxp_data_path_config = NULL,
 		.pxp_restart = pxp_software_restart,
-		.version = 32,
+		.version = PXP_V3_8ULP,
 	},
 };
 
@@ -1885,10 +1885,21 @@ static uint32_t pxp_fetch_shift_calc(uint32_t in_fmt, uint32_t out_fmt,
 
 static int pxp_start(struct pxps *pxp)
 {
-	__raw_writel(BM_PXP_CTRL_ENABLE_ROTATE1 | BM_PXP_CTRL_ENABLE |
-		BM_PXP_CTRL_ENABLE_CSC2 | BM_PXP_CTRL_ENABLE_LUT |
-		BM_PXP_CTRL_ENABLE_PS_AS_OUT | BM_PXP_CTRL_ENABLE_ROTATE0,
-			pxp->base + HW_PXP_CTRL_SET);
+	u32 val;
+
+	val = (BM_PXP_CTRL_ENABLE_ROTATE1 |
+	       BM_PXP_CTRL_ENABLE |
+	       BM_PXP_CTRL_ENABLE_CSC2 |
+	       BM_PXP_CTRL_ENABLE_PS_AS_OUT |
+	       BM_PXP_CTRL_ENABLE_ROTATE0 |
+	       BM_PXP_CTRL_BLOCK_SIZE);
+
+	if (pxp->devdata->version <= PXP_V3_8ULP) {
+		val |= BM_PXP_CTRL_ENABLE_LUT;
+		val &= ~(BM_PXP_CTRL_BLOCK_SIZE);
+	}
+
+	__raw_writel(val, pxp->base + HW_PXP_CTRL_SET);
 	dump_pxp_reg(pxp);
 
 	return 0;
-- 
2.25.1

